compiler/hash-ast-desugaring/src/lib.rs

```diff
@@ -17,49 +17,82 @@ use hash_ast::{
-                        .visit_expression(&(), expr.ast_ref_mut())
+use hash_pipeline::{sources::Sources, traits::Desugar, CompilerResult};
+use hash_source::{location::Span, SourceId};
+use std::{collections::HashSet, convert::Infallible};
+
+pub struct AstDesugaring;
+
+impl<'pool> Desugar<'pool> for AstDesugaring {
+    type State = HashSet<SourceId>;
+
+    fn make_state(&mut self) -> CompilerResult<Self::State> {
+        Ok(HashSet::default())
+    }
+
+    /// This function is used to lower all of the AST that is present within
+    /// the modules to be compatible with the typechecking stage. This is
+    /// essentially a pass that will transform the following structures
+    /// into a "simpler" variant:
+    ///
+    /// Any for-loops are transformed into a more simpler "loop" construct
+    /// with an inner match case that verifies that the iterator has no
+    /// more items that can be consumed.
+    ///
+    /// Any while-loops are also transformed into a simpler loop variant with
+    /// an inner match case that matches on the result of the while-loop
+    /// "condition" to see if it is "true" or "false". If it is false, then
+    /// the loop breaks, otherwise the body of the while-loop is executed.
+    ///
+    /// Any if-statements are transformed into equivalent match cases by using
+    /// the "if-guard" pattern to express all of the branches in the if-statement.
+    ///
+    /// This function utilised the pipeline thread pool in order to make the transformations
+    /// as parallel as possible. There is a queue that is queues all of the expressions within
+    /// each [hash_ast::ast::Module].
+    fn desugar(
+        &mut self,
+        target: SourceId,
+        sources: &mut Sources,
+        state: &mut Self::State,
+        pool: &'pool rayon::ThreadPool,
+    ) -> hash_pipeline::traits::CompilerResult<()> {
+        pool.scope(|scope| {
+            // De-sugar the target if it isn't already de-sugared
+            if !state.contains(&target) {
+                if let SourceId::Interactive(id) = target {
+                    let source = sources.get_interactive_block_mut(id);
+
+                    AstDesugaring
+                        .visit_body_block(&(), source.node_mut())
                         .unwrap()
-                })
+                }
             }
-        }
-    })
+
+            // Iterate over all of the modules and add the expressions
+            // to the queue so it can be distributed over the threads
+            for (id, module) in sources.iter_mut_modules() {
+                // Skip any modules that have already been de-sugared
+                if state.contains(&SourceId::Module(id)) {
+                    continue;
+                }
+
+                for expr in module.node_mut().contents.iter_mut() {
+                    scope.spawn(|_| {
+                        AstDesugaring
+                            .visit_expression(&(), expr.ast_ref_mut())
```

Comment:
Would be nice if the pool was a context to the visitor so that each expression could be queued rather than the top-level ones. Though I guess that might result in other bottlenecking related to queuing costs..

Reply:
Yes, that would be really nice, it could basically find body blocks and just queue all of the children to be "evaluated" and later processed.

I will add a note about this.